### DISCUSSION TITLE
Couldnt update table

### DISCUSSION THREAD

**Discussion by networkandcode** [supabase/supabase#3476] [Questions] [answered]
I have been trying to update a row using the following JS code
```
const { data, error } = await supabase
  .from('wrong_answers')
  .update({ wrong_answer_count: 2 })
  .eq('user_id', user.id);
```
I keep getting 404
user.id refers to the user auth id, and the table has row level policy, to allow alterations from that particular user.

console of the browser shows
```
browser-ponyfill.js?5d61:518 PATCH https://<app id>.supabase.co/rest/v1/wrong_answers?user_id=eq.<user-id> 404
```

The row that I am trying to modify, has the correct user_id

The SQL snippet ran for the table is as follows:
```
CREATE TABLE wrong_answers (
  id bigint generated by default as identity primary key,
  user_id uuid references auth.users not null,
  wrong_answer_count int,
  inserted_at timestamp with time zone default timezone('utc'::text, now()) not null
);

alter table wrong_answers enable row level security;

create policy "Individuals can insert their wrong answer count." on wrong_answers for
    insert with check (auth.uid() = user_id);

create policy "Individuals can update their wrong answer count." on wrong_answers for
    update with check (auth.uid() = user_id);

create policy "Wrong answers are public." on wrong_answers for
    select using (true);
```

**Comment by dshukertjr** [comment, marked as answer]
Hi @networkandcode !

I looked at your code, and am struggling to find the cause as well. I would double check again to see if there is actually a row with `user_id` that matches the logged in user's id again. Perhaps log the user id before you send the update request. Also while we're at it, let's see if we can get data with the same `eq` statement:

```typescript
console.log('user.id', user.id)

const { data: selectData, error: selectError } = await supabase
  .from('wrong_answers')
  .select()
  .eq('user_id', user.id);

console.log('selectData', selectData) // make sure you get the row at which you would like to update here

const { data, error } = await supabase
  .from('wrong_answers')
  .update({ wrong_answer_count: 2 })
  .eq('user_id', user.id);
```

Also, when you are writing row level security policies for updates, you should also include a `using` statement. `using` statement checks using the data that sits on the database before the update happening where as `with check` checks with data after the update if it were to happen. You need both to make your database secure!
```sql
create policy "Individuals can update their wrong answer count." on wrong_answers for
    update 
        using (auth.uid() = user_id)
        with check (auth.uid() = user_id);
```

**Reply by GaryAustin1** [reply]
Would not the update fail with RLS update policy without the USING?    I know update policy will use USING for a missing WITH CHECK, but I'm not sure the reverse is true...

**Reply by dshukertjr** [reply]
@GaryAustin1 
That must be it! Thanks for the followup!

**Reply by kevinknights29** [reply]
Thanks @dshukertjr, I was trying to create a policy based on the template provided in the UI. By adding the `using` portion of the code, I got it working 😄
<img width="1146" alt="image" src="https://github.com/supabase/supabase/assets/74464814/452edeb7-b116-417f-befb-893d3e3a79dc">

**Reply by M-Shahbaz** [reply]
I had a same issue, with new ui the Policy Behavior for read was set to Restrictive, changed to Permissive then update worked.

**Comment by networkandcode** [comment]
Thank you so much @dshukertjr It worked after the policy update.

**Comment by amirsinaa** [comment]
Hey everyone,
After I've upgraded to supabase-js@2 and ran [this security patch migration](https://github.com/supabase/supabase/discussions/9314) I can't update profiles table with RLS anymore

Update code: 

```typescript
const updates = {
        username,
        website,
        avatar_url
}

const { error } = await supabase.from('profiles').update(updates).eq('id', session.user.id).select()
```

RLS policy : 
<img width="1149" alt="Screen Shot 2022-10-17 at 4 03 27 PM" src="https://user-images.githubusercontent.com/20769213/196178093-33e5c9a2-8e1f-4cb3-a2f3-50f5bc4e4225.png">

Request sample : 
```
PATCH | https://mySiteID.supabase.co/rest/v1/profiles?id=eq.7a11ccd2-622d-4c87-a5e5-683f2639680c&select=*
```

Request headers : 
1PATCH /rest/v1/profiles?id=eq.7a11ccd2-622d-4c87-a5e5-683f2639680c&select=* HTTP/2
Host: mySiteID.supabase.co
User-Agent: my-device
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Referer: http://localhost:3000/
apikey: my-token
authorization: Bearer my-token
content-profile: public
content-type: application/json
prefer: return=representation
x-client-info: supabase-js/2.0.0
Content-Length: 106
Origin: http://localhost:3000
DNT: 1
Connection: keep-alive
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: cross-site
Sec-GPC: 1
TE: trailers
`
everything seems fine but I can't update the profiles 

P.S: I've disabled RLS just to make sure the problem originates from policies, not the code and it works fine without the RLS enabled

@GaryAustin1 
@dshukertjr

**Reply by GaryAustin1** [reply]
What is your select policy?   You must meet that also.  Otherwise check the authorization header and make sure it is not your anon key.

**Reply by amirsinaa** [reply]
> What is your select policy? You must meet that also.

My select policy is that profiles are viewable by everyone publicly. So the `USING expression` in the policies panel is set to one line of `true` boolean value

> check the authorization header and make sure it is not your anon key.

what should it be then? I've followed the documentation and created a supabase-client as below 

```
import { createClient } from '@supabase/supabase-js'

export const supabase = createClient(
  process.env.NEXT_PUBLIC_SUPABASE_URL,
  process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY
)
```

Isn't anon_key a jwt that can be used to access RLS tables?

**Reply by dshukertjr** [reply]
@amirsinaa 
Would it work if you include`id` in the request?

```typescript
const updates = {
        username,
        website,
        avatar_url
        id: session.user.id
}

const { error } = await supabase.from('profiles').update(updates).eq('id', session.user.id)
```

**Reply by amirsinaa** [reply]
@dshukertjr 
No luck, still getting a `404 Not Found` response.

**Reply by GaryAustin1** [reply]
Do you have a role in the RLS?  In the TO part or set with the dashboard UI drop down.

**Reply by dshukertjr** [reply]
@amirsinaa Feel free to contact support@supabase.com if the problem persists!

**Comment by renellc** [comment]
Hey all. I just encountered the issue @amirsinaa came across as well when I tried to update my table. My table had a single RLS policy on UPDATE, which looked like the following:

![image](https://github.com/supabase/supabase/assets/25127718/85480d9c-b4ae-4e4f-a4c0-a86652b7a629)

This is how I was calling it within a route handler in NextJS 13+:

```typescript
const supabase = createRouteHandlerClient<Database>({ cookies });

const { data: { user }, error: authError } = await supabase.auth.getUser();
if (authError) {
  return NextResponse.json({ error: "Could not update username" }, { status: authError.status });
}

const { data, error, status } = await supabase
  .from("profiles")
  .update({ username: "NEW_USERNAME", completed_signup: true })
  .eq("id", user?.id)
  .select();
```

I was able to fix this issue by adding another RLS policy allowing users to read from the table as well. It can either be allowing users to just read their own rows, or allowing users to read all rows:

![image](https://github.com/supabase/supabase/assets/25127718/7fa065b2-f12e-442c-80e8-30e99a41f641)

![image](https://github.com/supabase/supabase/assets/25127718/a9cfc849-1c73-4dd3-8614-6bca29aca78a)

Adding either of the above policies finally let me update a user's row in my table.

@GaryAustin1 @dshukertjr I'm not entirely sure if this is intended behavior or not as it would be odd to me that I can't have just a single UPDATE policy without having a SELECT policy as well (not sure if this is a bug or if it's a Postgres thing). I thought I should leave a comment on how I solved this particular issue for future readers, and maybe in the case that this issue hasn't been resolved yet.

**Comment by danya0365** [comment]
client just can only read but can not update own data, the api always return 403

this tutorial is not working https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs

**Reply by GaryAustin1** [reply]
I suggest you start a new discussion with more details on your RLS and failing call.  Another way to get help is to use the Supabase Discord forum.   At any rate sounds like you don't have RLS set up properly for update, or the place you are doing the update from has a client that does not have a user session.

**Reply by danya0365** [reply]
I try to read data before update 

```
import { createClientComponentClient } from '@supabase/auth-helpers-nextjs';

const supabase = createClientComponentClient();

const { data: userDetails } = await supabase
      .from('users')
      .select('*')
      .single();

```

then try to update but api return 403

```
let { error } = await supabase.from('users').upsert({
        id: userDetails!.id,
        avatar_url: newAvatarUrl
      });
      if (error) throw error;
```

this is 100% problem is on my RLS

i will try to fix this by my self

```code: '42501', details: null, hint: null, message: 'new row violates row-level security policy for table "users"```
